File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Integer> stack = new Stack<>();

        for(int i=0; i<num.length(); i++){
            while(!stack.isEmpty() && k>0 && num.charAt(i) < num.charAt(stack.peek())){
                stack.pop();
                k--;
            }

            stack.push(i);
        }

        while(k > 0){
                stack.pop();
                k--;
        }

        StringBuilder str = new StringBuilder();
        while(!stack.isEmpty()){
            int index = stack.pop();
            str.append(num.charAt(index));
        }

        str.reverse();

        StringBuilder result = new StringBuilder();
        boolean foundNonZero = false;

        for (int i=0; i<str.length(); i++) {
            char ch = str.charAt(i);

            if (ch != '0') {
                foundNonZero = true;
            }

            if (foundNonZero) {
                result.append(ch);
            }
        }

        // If everything was zero, return "0"
        if (result.length() == 0) {
            return "0";
        }

        return result.toString();
    }
}

File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        stack<int> st;

        // Store indices in the stack
        for (int i = 0; i < num.length(); i++) {

            while (!st.empty() &&
                   k > 0 &&
                   num[i] < num[st.top()]) {

                st.pop();
                k--;
            }

            st.push(i);
        }

        // If k is still remaining, remove from the end
        while (k > 0 && !st.empty()) {
            st.pop();
            k--;
        }

        // Build the number from remaining indices
        string str = "";

        while (!st.empty()) {
            int index = st.top();
            st.pop();

            str += num[index];
        }

        // Stack gives us the digits backwards
        reverse(str.begin(), str.end());

        // Remove leading zeroes
        string result = "";
        bool foundNonZero = false;

        for (int i = 0; i < str.length(); i++) {

            if (str[i] != '0') {
                foundNonZero = true;
            }

            if (foundNonZero) {
                result += str[i];
            }
        }

        // If nothing remains, answer is "0"
        if (result.empty()) {
            return "0";
        }

        return result;
    }
};